### PR TITLE
Collision detection + SFML display

### PR DESCRIPTION
**Pull request by alexdewar**
Hey,

This is just something I've been working on for the robot experiments we were going to do in the drone zone.

I've added a simple 2D collision detector. You input the dimensions of the robot and of the objects you're avoiding + how close the robot is allowed to get, then you can check if a collision has occurred or not.

I've also included a little generic SFML display for viewing (you might need to install the libs on your machine).

### CODE REVIEW

**Comment by neworderofjamie** [comment]
I'm trying to do a little bit of GeNN coding this afternoon so reviewing may have to wait until next week. However, my immediate thought is do we really want bits of our code using SFML, GLFW and SDL? 

- All three let you create windows with OpenGL contexts in a cross-platform way. 
- GLFW does nothing else and is pleasantly modern (which, as that was what I wanted, is why I chose it originally)
- SDL does some other stuff and is decidedly elderly, written and C, but would let us target important platforms like Amiga...
-  SFML does a whole load of stuff and looks perfectly nice.

I think the ability to do stuff like this P.R. is probably a good thing so maybe we should port everything across to SFML? Let me know your thoughts and, if you agree, we can split that thankless task between us.

**Comment by alexdewar** [comment]
I'm all for using SFML -- it's got a really nice interface. The only problem is that you can only do 2D stuff with it, so we couldn't use it for e.g. libantworld :disappointed: 

Norbert's ``CarDisplay`` class could be replaced with the ``SFMLRenderer`` though.

**Comment by neworderofjamie** [comment]
According to https://www.sfml-dev.org/tutorials/2.5/window-opengl.php that would seem not to be the case!

**Comment by alexdewar** [comment]
Ahh sorry, I see what you meant now. Yes, I guess we could drop the GLFW stuff and mix and match SFML with OpenGL.

That's not a bad idea. Replacing the GLFW bits with SFML would probably make the code more readable apart from anything.

I'm happy to divide this task up!

**Comment by neworderofjamie** [comment]
Hmm I think that could be a little optimistic - all the OpenGL state
fiddling will remain cos that's the nature of the beast beneath but I think
sticking with one library around the outside is a good thing

**Comment by alexdewar** [comment]
Hey,

Would you mind having a quick look at this when you get a chance? It's not urgent, but I've got a couple more branches based on this (with pretty minor stuff in) that I'd also like to merge soonish.

**Comment by alexdewar** [comment]
PS -- Not sure if you have SFML installed or not? Looks like it can't find the package.

**Comment by alexdewar** [comment]
Actually, I've just realised it makes sense to merge some later stuff into this branch. Hang on a mo

**Comment by alexdewar** [comment]
Ok, done. I also added the functionality we have with the SDL code into ``SFMLDisplay``. Originally I also deleted all the SDL stuff on this branch, but I've reverted that as it looks like @tenxlenx is still using it.

**Comment by tenxlenx** [comment]
I use SDL because for some mysterious reason SFML does not work on the office computer, otherwise I prefer SFML because it has so much more functionality

**Comment by neworderofjamie** [comment]
@alexdewar and I are good at breaking office computers - let us know when you're next in and we can fight it.

File: common/sfml_display.h [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+// Standard C++ includes
+#include <vector>
+#include <stdexcept>
+
+namespace BoBRobotics {
+using namespace units::literals;
+
+template<typename LengthUnit = units::length::meter_t>
+class SFMLDisplay
```
Comment (neworderofjamie) [inline common/sfml_display.h]: Maybe ``SFMLAgentArena`` or ``SFMLWorld`` would be better names - display suggests to me at least that it's a lower-level thing

File: common/pose.h [hunk 2]
```diff
@@ -25,6 +29,8 @@ template<typename Derived>
 class PoseBase
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply common/pose.h]: Actually I am using a CRTP here... or is that not what you meant?

[the remark this answers is not in the document]
Comment (neworderofjamie) [reply common/pose.h]: Ahh ok, in that case ``Derived`` is probably an alright name

File: common/collision.h [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        for (size_t i = 0; i < vectors.size(); i++) {
+            matrix(i, 0) = static_cast<meter_t>(vectors[i].x()).value();
+            matrix(i, 1) = static_cast<meter_t>(vectors[i].y()).value();
+        }
+        return matrix;
+    }
+
+    static constexpr meter_t inf()
+    {
+        return meter_t{ std::numeric_limits<double>::infinity() };
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply common/collision.h]: This doesn't compile for me. I've used the ``UNIT_LIB_DEFAULT_TYPE`` macro instead, which lets you get (and set) the unit lib type.

File: common/sfml_display.h (hunk 1 above)
Comment (alexdewar) [reply common/sfml_display.h]: :+1:

**Comment by alexdewar** [comment]
Ta for looking through! Would you mind installing SFML so that we can pass the unit tests?

File: common/collision.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (neworderofjamie) [reply common/collision.h]: maybe you need an extra ``typedef `` or some shit...

**Pull request merged**